File: python/lombokalgoritma/graph/dfs.py
# LombokAlgoritma — depth-first search (SPEC §9.2)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from .types import Graph, assert_node, build_adj_list


def dfs(g: Graph, source: int) -> list[int]:
    """Iterative pre-order from ``source``; unvisited neighbours pushed in reverse edge order."""
    adj = build_adj_list(g)
    assert_node(g, source, "source")
    visited = [False] * len(adj)
    order: list[int] = []
    stack = [source]
    while stack:
        u = stack.pop()
        if visited[u]:
            continue
        visited[u] = True
        order.append(u)
        for v, _ in reversed(adj[u]):
            if not visited[v]:
                stack.append(v)
    return order

File: rust/lombokalgoritma/src/graph/dfs.rs
// LombokAlgoritma — depth-first search (SPEC §9.2)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use super::types::{adjacency, check_node, Graph};
use crate::Result;
use alloc::vec;
use alloc::vec::Vec;

/// Pre-order of an iterative DFS from `source`: pop a node, skip it if visited, otherwise emit it
/// and push its unvisited neighbours in **reverse** edge-list order. O(V + E).
///
/// # Errors
/// [`crate::Error::OutOfRange`] for an edge endpoint or `source` outside `[0, n)`.
pub fn dfs(g: &Graph, source: usize) -> Result<Vec<usize>> {
    let adj = adjacency(g)?;
    check_node(g, source)?;
    let mut visited = vec![false; g.nodes];
    let mut order = Vec::new();
    let mut stack = vec![source];
    while let Some(u) = stack.pop() {
        if visited[u] {
            continue;
        }
        visited[u] = true;
        order.push(u);
        for &(to, _) in adj[u].iter().rev() {
            if !visited[to] {
                stack.push(to);
            }
        }
    }
    Ok(order)
}
